SpatialGDK/Source/SpatialGDK/Private/Utils/MetricsExport.cpp

```diff
@@ -0,0 +1,161 @@
+	}
+	FString ProjectNameFromCli;
+	if (FParse::Value(FCommandLine::Get(), TEXT("appName"), ProjectNameFromCli))
+	{
+		ProjectName = ProjectNameFromCli;
+	}
+	else
+	{
+		ProjectName = TEXT("local_project");
+	}
+	FString DeploymentNameFromCli;
+	if (FParse::Value(FCommandLine::Get(), TEXT("deploymentName"), DeploymentNameFromCli))
+	{
+		DeploymentName = DeploymentNameFromCli;
+	}
+	else
+	{
+		DeploymentName = TEXT("local_dpl");
+	}
+}
+
+bool UMetricsExport::ShouldCreateSubsystem(UObject* Outer) const
+{
+	return GetDefault<USpatialGDKSettings>()->bEnableMetricsExport;
+}
+
+void UMetricsExport::Initialize(FSubsystemCollectionBase& Collection)
+{
+	Super::Initialize(Collection);
+
+	USpatialGameInstance* SpatialGameInstance = Cast<USpatialGameInstance>(GetGameInstance());
+	if (SpatialGameInstance == nullptr)
+	{
+		return;
+	}
+
+	SpatialGameInstance->OnSpatialConnected.AddUniqueDynamic(this, &UMetricsExport::SpatialConnected);
+}
+
+bool UMetricsExport::Tick(float DeltaSeconds)
+{
+	FramesSinceLastFpsWrite++;
+	const int64 Now = FDateTime::Now().ToUnixTimestamp();
+	const int64 TimeSinceLastUpdate = Now - LastUpdateSendTime;
+	if (TimeSinceLastUpdate > InfluxMetricsSendGapSecs)
+	{
+		const uint32 Fps = FramesSinceLastFpsWrite / (Now - LastUpdateSendTime);
+		WriteMetricsToProtocolBuffer(WorkerId, TEXT("fps"), Fps);
+		FlushProtocolBufferToDatabase();
+		LastUpdateSendTime = Now;
+		FramesSinceLastFpsWrite = 0;
+	}
+
+	return true;
+}
+
+void UMetricsExport::SpatialConnected()
+{
+	USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(GetWorld()->GetNetDriver());
+	if (NetDriver == nullptr)
+	{
+		if (UPendingNetGame* PendingNetGame = GetGameInstance()->GetWorldContext()->PendingNetGame)
+		{
+			NetDriver = Cast<USpatialNetDriver>(PendingNetGame->GetNetDriver());
+		}
+	}
+
+	if (NetDriver == nullptr)
+	{
+		return;
+	}
+
+	WorkerType = NetDriver->IsServer() ? TEXT("Server") : TEXT("Client");
+	WorkerId = NetDriver->Connection->GetWorkerId();
+
+	if (!TickerHandle.IsValid())
+	{
+		TickerHandle = FTicker::GetCoreTicker().AddTicker(FTickerDelegate::CreateUObject(this, &UMetricsExport::Tick));
+	}
+}
+
+void UMetricsExport::WriteMetricsToProtocolBuffer(const FString& Worker, FString Field, float Value)
+{
+	MetricsMap.FindOrAdd(Worker).FindOrAdd(Field) = Value;
+}
+
+void UMetricsExport::FlushProtocolBufferToDatabase()
+{
+	for (const auto& WorkerMetrics : MetricsMap)
+	{
+		FString Lines = "";
+
+		const FString& WorkerIdentifier = WorkerMetrics.Key;
+		Lines.Append(FString::Printf(TEXT("worker_data,worker=%s,project=%s,deployment=%s,type=%s "), *WorkerIdentifier, *ProjectName,
+									 *DeploymentName, *WorkerType));
+		const TMap<FString, float>& Metrics = WorkerMetrics.Value;
+		for (const auto& MetricKeyValue : Metrics)
+		{
+			Lines.Append(FString::Printf(TEXT("%s=%f,"), *MetricKeyValue.Key, MetricKeyValue.Value));
+		}
+		Lines.RemoveFromEnd(TEXT(","));
+		PostLineProtocolToInfluxDBServer(Lines);
```

Comment:
We send an http request for each metric. Could we send all metrics in "Lines" one string?

Reply:
If this becomes a performance issue we can investigate 👍